<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Accurate Tiling of Spatial Single-Cell Data â€¢ tessera</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Accurate Tiling of Spatial Single-Cell Data">
<meta name="description" content="An algorithm for segmenting single-cell resolution spatial omics data into small multicellular tiles whose edges track with natural tissue boundaries. These tiles can then be used in downstream analysis to label and define tissue regions across samples.">
<meta property="og:description" content="An algorithm for segmenting single-cell resolution spatial omics data into small multicellular tiles whose edges track with natural tissue boundaries. These tiles can then be used in downstream analysis to label and define tissue regions across samples.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">tessera</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/vignette_basic.html">Quickstart: Using Tessera on a Single Sample</a></li>
    <li><a class="dropdown-item" href="articles/vignette_stepthrough.html">Walkthrough: Tessera Algorithm Step-by-Step</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/korsunskylab/tessera/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="tessera">Tessera<a class="anchor" aria-label="anchor" href="#tessera"></a>
</h1></div>
<p><em>Accurate tiling of spatial single-cell data with Tessera</em></p>
<p><strong>Tessera</strong> is an algorithm for segmenting single-cell resolution spatial omics data into small multicellular tiles whose edges track with natural tissue boundaries. These tiles can then be used in downstream analysis to label and define tissue regions across samples.</p>
<p>Check out the manuscript on <a href="https://www.biorxiv.org/content/10.1101/2025.01.17.633630v1" class="external-link">bioRxiv</a> for additional details.</p>
<p><img src="reference/figures/cartoon.png"></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The Tessera algorithm takes as input single cells (or pixels) with spatial coordinates and cell embeddings (or transcript counts) for each cell. The output is a segmentation of adjacent cells into tiles with a user-controlled size parameter. Boundaries between tiles align with where cell composition and gene expression change the most within the tissue. Segmentation using the Tessera algorithm has four main steps:</p>
<ol style="list-style-type: decimal">
<li>
<em>Constructing inputs:</em> A triangle mesh is constructed using Delauney triangulation and pruned to eliminate long edges. If transcript counts are provided for each cell instead of embeddings, then cell embeddings are computed using principal component analysis (PCA).</li>
<li>
<em>Gradient estimation:</em> Gradients are calculated at each vertex by considering the difference in cell embeddings between each cell and its neighbors in the mesh. These gradients are smoothed using anisotropic bilateral filtering, and then gradients are defined for edges and triangles in the mesh by averaging the vertices that each edge or triangle contains.</li>
<li>
<em>Tissue segmentation using discrete Morse theory (DMT):</em> A scalar field is defined by taking the magnitude of the total gradient at each vertex, edge, and triangle. Then DMT-based segmentation is performed by constructing a maximum spanning forest on the triangles and a minimum spanning forest on the vertices. Separatrices that partition cells into tiles of homogeneous composition are defined by tracing paths between critical points, specifically between saddle edges and maximum triangles.</li>
<li>
<em>Hierarchical agglomeration:</em> Tiles from DMT-based segmentation are merged using single-linkage agglomerative clustering to obtain tiles containing a number of cells between a user-provided minimum and maximum value. Pairs of adjacent tiles are scored according to their transcriptional similarity, compactness of shape after merging, and number of cells, in order to prioritize favorable merges in each agglomerative clustering step.</li>
</ol>
</div>
<div class="section level2">
<h2 id="system-requirements">System Requirements<a class="anchor" aria-label="anchor" href="#system-requirements"></a>
</h2>
<div class="section level3">
<h3 id="os-requirements">OS Requirements<a class="anchor" aria-label="anchor" href="#os-requirements"></a>
</h3>
<p><code>tessera</code> is supperted for <em>macOS</em> and <em>Linux</em>. The package has been tested on the following systems:</p>
<ul>
<li>macOS: Sonoma (14.6.1)</li>
<li>Linux: CentOS 7 (7.9.2009)</li>
</ul>
</div>
<div class="section level3">
<h3 id="r-dependencies">R Dependencies<a class="anchor" aria-label="anchor" href="#r-dependencies"></a>
</h3>
<p><code>tessera</code> has been tested on R versions &gt;= 4.3. Please consult the DESCRIPTION file for more details on required R packages, including Rcpp for R and C++ integraion.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="install-from-github">Install from GitHub<a class="anchor" aria-label="anchor" href="#install-from-github"></a>
</h3>
<p>Open R and run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">'korsunskylab/tessera'</span><span class="op">)</span></span></code></pre></div>
<p>This will install dependencies from CRAN, which can be slow and fails on some systems. Instead, it is recommended to install the dependencies manually using conda/mamba (see below).</p>
</div>
<div class="section level3">
<h3 id="manually-install-dependencies-with-condamamba-recommended">Manually install dependencies with conda/mamba (recommended)<a class="anchor" aria-label="anchor" href="#manually-install-dependencies-with-condamamba-recommended"></a>
</h3>
<p>In the command line:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> tessera_env</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">mamba</span> activate tessera_env</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">## If using macOS with Apple Silicon:</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co"># conda config --env --set subdir osx-64</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co"># Install required dependencies (~1 min)</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-essentials r-rcpp r-rcpparmadillo r-bh r-devtools <span class="dt">\</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>r-tidyverse r-matrix r-rlang r-r.utils r-sf r-igraph r-furrr r-future r-data.table <span class="dt">\</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>r-geometry r-mclust r-rspectra r-magrittr r-harmony</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co"># Optionally install Seurat and additional packages to run vignettes (~15 sec)</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">-c</span> conda-forge r-seurat r-ggthemes r-patchwork r-viridis jupyterlab r-irkernel</span></code></pre></div>
<p>Next, in an R console:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">'korsunskylab/tessera'</span>, dependencies <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>  <span class="co"># ~1 min</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<div class="section level3">
<h3 id="standalone-mode">Standalone Mode<a class="anchor" aria-label="anchor" href="#standalone-mode"></a>
</h3>
<p>Check out <code><a href="articles/vignette_basic.html">vignette("vignette_basic")</a></code> for a quick start tutorial which demonstrates using <code>tessera</code> in standalone mode on a single sample.</p>
<p>The basic usage is as follows:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="reference/GetTiles.html">GetTiles</a></span><span class="op">(</span></span>
<span>    X <span class="op">=</span> <span class="va">meta_data</span><span class="op">$</span><span class="va">X</span>,         <span class="co"># Vector of cell spatial coordinates</span></span>
<span>    Y <span class="op">=</span> <span class="va">meta_data</span><span class="op">$</span><span class="va">Y</span>,         <span class="co"># Vector of cell spatial coordinates</span></span>
<span>    counts <span class="op">=</span> <span class="va">counts</span>,         <span class="co"># Gene-by-cell matrix of transcript counts</span></span>
<span></span>
<span>    embeddings <span class="op">=</span> <span class="va">embeddings</span>, <span class="co"># (Optional) Cell-by-embedding matrix of pre-computed cell embeddings. If missing, embeddings are calculated using PCA.</span></span>
<span>    meta_data <span class="op">=</span> <span class="va">meta_data</span>,   <span class="co"># (Optional) Additional cell meta data</span></span>
<span>    meta_vars_include <span class="op">=</span> <span class="va">meta_vars_include</span>, <span class="co"># (Optional) Cell meta data to include in output</span></span>
<span></span>
<span>    group.by <span class="op">=</span> <span class="st">'sample_id'</span>,  <span class="co"># (Optional) Name of meta_data column that provides sample IDs. If missing, treated as a single sample.</span></span>
<span></span>
<span>    <span class="co"># Additional Tessera algorithm parameters</span></span>
<span>    prune_thresh_quantile <span class="op">=</span> <span class="fl">0.99</span>, prune_min_cells <span class="op">=</span> <span class="fl">1</span>, <span class="co"># Control pruning of long edges and disconnected cells</span></span>
<span>    max_npts <span class="op">=</span> <span class="fl">50</span>, min_npts <span class="op">=</span> <span class="fl">5</span>,                       <span class="co"># Control size of Tessera tiles</span></span>
<span>    <span class="va">...</span>                      </span>
<span><span class="op">)</span></span>
<span><span class="va">dmt</span> <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">dmt</span>                <span class="co"># Mesh data structures with results from segmentation</span></span>
<span><span class="va">aggs</span> <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">aggs</span>              <span class="co"># Tiles resulting from DMT-based segmentation and agglomeration</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="seurat-objects-multi-sample">Seurat Objects (Multi-sample)<a class="anchor" aria-label="anchor" href="#seurat-objects-multi-sample"></a>
</h3>
<p>Tessera can also be applied directly to a Seurat object containing single cells with spatial coordinates. The <code>GetTiles</code> function can use cell embeddings that have already been pre-computed (and integrated, if there are multiple samples). By default, the output is a pair of Seurat objects: 1) a single-cell Seurat object updated with tile assignments for each cell, and 2) a Seurat object where each entry represents an individual Tessera tile.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize<span class="op">=</span> <span class="fl">4</span><span class="op">*</span><span class="fl">1024</span><span class="op">^</span><span class="fl">3</span><span class="op">)</span>   <span class="co"># For larger datasets, the memory allowance may need to be increased for parallelization</span></span>
<span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multicore.html" class="external-link">multicore</a></span><span class="op">)</span>             <span class="co"># Parallelize over multiple samples (if doing multi-sample analysis)</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="reference/GetTiles.html">GetTiles</a></span><span class="op">(</span></span>
<span>    <span class="va">obj</span>,        <span class="co"># Single-cell Seurat object</span></span>
<span>    <span class="st">'spatial'</span>,  <span class="co"># Name of dimesional reduction where x/y coordinates are stored</span></span>
<span></span>
<span>    embeddings <span class="op">=</span> <span class="st">'harmony'</span>,  <span class="co"># (Optional) Name of dimensional reduction where pre-computed single-cell embeddings are stored</span></span>
<span>    group.by <span class="op">=</span> <span class="st">'sample_id'</span>,  <span class="co"># (Optional) Name of meta.data column that provides sample IDs. If missing, treated as a single sample.</span></span>
<span></span>
<span>    <span class="co"># Additional Tessera algorithm parameters</span></span>
<span>    prune_thresh_quantile <span class="op">=</span> <span class="fl">0.99</span>, prune_min_cells <span class="op">=</span> <span class="fl">1</span>, <span class="co"># Control pruning of long edges and disconnected cells</span></span>
<span>    max_npts <span class="op">=</span> <span class="fl">50</span>, min_npts <span class="op">=</span> <span class="fl">5</span>,                       <span class="co"># Control size of Tessera tiles</span></span>
<span>    <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span><span class="va">obj</span> <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">obj</span>                <span class="co"># Seurat object of single-cells (with cell-to-tile mapping)</span></span>
<span><span class="va">tile_obj</span> <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">tile_obj</span>      <span class="co"># Seurat object of Tessera tiles</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="vignettes">Vignettes:<a class="anchor" aria-label="anchor" href="#vignettes"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Quickstart (approx. runtime: &lt;10 sec): <code><a href="articles/vignette_basic.html">vignette("vignette_basic")</a></code>
</li>
</ol>
<p><a href="https://github.com/korsunskylab/tessera/blob/main/vignettes/vignette_basic.ipynb" class="external-link uri">https://github.com/korsunskylab/tessera/blob/main/vignettes/vignette_basic.ipynb</a></p>
<ol start="2" style="list-style-type: decimal">
<li>Walkthrough (approx. runtime: &lt;10 sec): <code><a href="articles/vignette_stepthrough.html">vignette("vignette_stepthrough")</a></code>
</li>
</ol>
<p><a href="https://github.com/korsunskylab/tessera/blob/main/vignettes/vignette_stepthrough.ipynb" class="external-link uri">https://github.com/korsunskylab/tessera/blob/main/vignettes/vignette_stepthrough.ipynb</a></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/korsunskylab/tessera/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/korsunskylab/tessera/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tessera</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Ilya Korsunsky <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0003-4848-3948" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Daniel Stein <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-9442-2604" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ilya Korsunsky, Daniel Stein.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
